### PR TITLE
Reflectometry GUI restructuring: Plotting

### LINKED ISSUE #25431: ISIS Reflectometry GUI plotting capability
As part of the port to Workbench these features need to be added first. 
Please refer to issue #25383 to see tests for expected functionality.

### PR DESCRIPTION
**Pull request by Pasarus**
**Description of work.**
The new restructure of the GUI does not have plotting functionality. As this is a major part of the Porting of the GUI to Workbench, I have added it ahead of porting it to workbench.

**To test:**

**Note for testsing, the data will look different but that is not the fault of this PR**

### First, set up some processed rows
- Open the `ISIS Reflectometry` interface, select `INTER`.
- From the Reflectometry menu, select Import table and choose `INTER_NR_test2.tbl` from the ISIS sample data.
- Click `Process` and when finished expand all groups. All rows should now be green.

### Test the first button ("plot runs")
- Select a single row within a group and then click plot runs. You get a single curve:

- Select two rows and click plot runs. You get two curves on one plot:

- Select a group and click plot runs. You get all of the curves from the group on one plot:

- Select one row from the first group and two rows from the second group. You should get three curves:

- Select one row from the first group and the overall group for the second group. You should get three curves:

### Test the second button ("plot group")
- Select a row or rows within a group and click plot group. Nothing should happen (ideally this should be disabled).

- Select a group and click plot group. You should get a single curve (for the stitched result):

- Select two groups and click plot group. You should get two curves:

- Select one row from the first group and the overall group for the second group. The row should be ignored and one curve plotted for the second group:

*Tests posted by @gemmaguest on the issue*

Fixes #25431 . 

*This does not need release notes as I believe it is an expected feature of the new GUI already*

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by gemmaguest — CHANGES_REQUESTED** [review]
I think there is a better approach than passing the notifications up through the views. Sorry I didn't think of this before! We could use an interface similar to `IMessageHandler`:
- Add an interface `IPlotter` with a pure virtual method `reflectometryPlot()` (for ease of testing and extendibility).
- Inherit this from your `Plotter` class (this could perhaps be split into 2 separate classes for the Qt4 and Qt5 stuff to avoid having repeated checks for the Qt version. What you have is probably fine for now, except that it might be easier to test if split up.)
- The `Plotter` class for Qt4 should have a pointer to the `IMainWindowView` rather than the `RunsTableView`. This already has a method `runPythonAlgorithm`. The other views then shouldn't need to know about the python stuff.
- Set up the `IPlotter` for the relevant version of Qt in `MainWindowView::initLayout`, and pass it to the `RunsTablePresenterFactory` and through to the `RunsTablePresenter`.
- The `RunsTablePresenter` should then just be able to call `IPlotter::reflectometryPlot()`.

What do you think?

Otherwise:
- code quality looks good and works well :+1:
- some minor comments in-line
- New unit tests should be added. Having an `IPlotter` interface you can mock should help when testing `RunsTablePresenter`, and to test the Qt4 `Plotter` you could use a `MockMainWindowView`

File: qt/scientific_interfaces/ISISReflectometry/GUI/RunsTable/RunsTablePresenter.cpp [hunk 1]
```diff
@@ -565,5 +565,51 @@ bool RunsTablePresenter::isProcessing() const {
...
+
+#if QT_VERSION < QT_VERSION_CHECK(5, 0, 0)
+  const auto runsTable = dynamic_cast<RunsTableView *>(m_view);
+  Plotter plotter(runsTable);
+#else
+  Plotter plotter;
+#endif
+  plotter.reflectometryPlot(workspaces);
+
+  m_view->jobs().clearSelection();
```
[the remark this answers is not in the document]
Comment (Pasarus) [reply qt/scientific_interfaces/ISISReflectometry/GUI/RunsTable/RunsTablePresenter.cpp]: That was the behavior of the previous GUI, I could always remove this?

File: qt/scientific_interfaces/ISISReflectometry/GUI/RunsTable/RunsTablePresenter.cpp [hunk 2]
```diff
@@ -565,5 +565,51 @@ bool RunsTablePresenter::isProcessing() const {
...
   return m_mainPresenter->isAutoreducing();
 }
+
+void RunsTablePresenter::notifyPlotSelectedPressed() {
+  std::vector<std::string> workspaces;
+  const auto rows = m_model.selectedRows();
+
+  for (const auto &row : rows) {
+    const auto workspaceName = row.reducedWorkspaceNames().iVsQBinned();
+    if (workspaceName != "")
```
[the remark this answers is not in the document]
Comment (Pasarus) [reply qt/scientific_interfaces/ISISReflectometry/GUI/RunsTable/RunsTablePresenter.cpp]: :+1:

File: qt/scientific_interfaces/ISISReflectometry/GUI/Runs/RunsView.cpp [hunk 3]
```diff
@@ -5,6 +5,7 @@
 //     & Institut Laue - Langevin
 // SPDX - License - Identifier: GPL - 3.0 +
 #include "RunsView.h"
+#include "../Batch/BatchView.h"
```
Comment (gemmaguest) [inline qt/scientific_interfaces/ISISReflectometry/GUI/Runs/RunsView.cpp]: Is this needed?

File: qt/scientific_interfaces/ISISReflectometry/GUI/Runs/RunsView.h [hunk 4]
```diff
@@ -123,6 +124,7 @@ class MANTIDQT_ISISREFLECTOMETRY_DLL RunsView
   QBasicTimer m_timer;
 
   RunsTableView *m_tableView;
+  BatchView *m_batchView;
```
Comment (gemmaguest) [inline qt/scientific_interfaces/ISISReflectometry/GUI/Runs/RunsView.h]: I don't think this is used any more? This and the forward declaration could be removed.

**Review by gemmaguest — CHANGES_REQUESTED** [review]
Thanks for the changes - I think this is much easier to follow :+1:

Just some comments regarding minor things to tidy up - mainly include statements which I think have been left in accidentally, and some uses of pointers where I think references would do.

File: qt/scientific_interfaces/ISISReflectometry/GUI/Batch/BatchView.h [hunk 5]
```diff
@@ -39,6 +41,7 @@ class BatchView : public QWidget, public IBatchView {
       std::deque<MantidQt::API::IConfiguredAlgorithm_sptr> algorithms) override;
   void executeAlgorithmQueue() override;
   void cancelAlgorithmQueue() override;
+  void executePythonCode(const std::string &pythonCode);
```
Comment (Pasarus) [inline qt/scientific_interfaces/ISISReflectometry/GUI/Batch/BatchView.h]: Is this needed?

File: qt/scientific_interfaces/ISISReflectometry/GUI/RunsTable/RunsTablePresenter.cpp [hunk 6]
```diff
@@ -6,18 +6,18 @@
 // SPDX - License - Identifier: GPL - 3.0 +
 #include "RunsTablePresenter.h"
 #include "Common/Map.h"
+#include "GUI/Plotting/Plotter.h"
```
[the remark this answers is not in the document]
Comment (Pasarus) [reply qt/scientific_interfaces/ISISReflectometry/GUI/RunsTable/RunsTablePresenter.cpp]: It should be IPlotter.h

File: qt/scientific_interfaces/ISISReflectometry/Reduction/ReductionJobs.cpp [hunk 7]
```diff
@@ -8,6 +8,7 @@
 #include "Common/IndexOf.h"
 #include "Common/Map.h"
 #include "MantidQtWidgets/Common/Batch/AssertOrThrow.h"
+#include "RowLocation.h"
 #include <iostream>
```
Comment (Pasarus) [inline qt/scientific_interfaces/ISISReflectometry/Reduction/ReductionJobs.cpp]: Remove iostream?

File: qt/widgets/common/inc/MantidQtWidgets/Common/Batch/JobTreeView.h [hunk 8]
```diff
@@ -33,6 +33,7 @@ class EXPORT_OPT_MANTIDQT_COMMON JobTreeView : public QTreeView,
 public:
   JobTreeView(QStringList const &columnHeadings, Cell const &defaultCellStyle,
               QWidget *parent = nullptr);
+  ~JobTreeView(){};
```
Comment (Pasarus) [inline qt/widgets/common/inc/MantidQtWidgets/Common/Batch/JobTreeView.h]: Nessercary?

**Comment by Pasarus** [comment]
@gemmaguest I believe I have fixed all of the associated fixes/requests on this PR. Could you give it another review please?

**Review by gemmaguest — APPROVED** [review]
This all looks good now and all combinations of plots work :+1: :+1:

Some notes from testing:
- The .tbl file can't be imported in the restructured GUI yet so you have to enter the run numbers and angles separately.
- As noted in the test instructions, the differences in plots are not related to this PR (I get the same differences on the feature branch we're merging into). I'll look into that in a future PR. The main thing to test here is that the correct workspaces are being plotted, which they are.
- There is a crash when processing all groups which is also not related to this PR as I can reproduce it in the feature branch we're merging into. I'll look into that in a future PR.

**Pull request merged**